src/main/java/appeng/tile/inventory/AppEngInternalAEInventory.java

```diff
@@ -150,42 +142,85 @@ public ItemStack getStackInSlot( final int var1 )
 	}
 
 	@Override
-	public ItemStack decrStackSize( final int slot, final int qty )
+	public ItemStack insertItem( int slot, ItemStack stack, boolean simulate )
 	{
-		if( this.inv[slot] != null )
+		if( stack.isEmpty() )
 		{
-			final ItemStack split = this.getStackInSlot( slot );
-			ItemStack ns = ItemStack.EMPTY;
+			return ItemStack.EMPTY;
+		}
+
+		ItemStack existing = getStackInSlot( slot );
+		int limit = getStackLimit( slot, stack );
+
+		if( !existing.isEmpty() )
+		{
+			if( !ItemHandlerHelper.canItemStacksStack( stack, existing ) )
+				return stack;
 
-			if( qty >= split.getCount() )
+			limit -= existing.getCount();
+		}
+
+		if( limit <= 0 )
+			return stack;
+
+		boolean reachedLimit = stack.getCount() > limit;
+
+		if( !simulate )
+		{
+			if( existing.isEmpty() )
 			{
-				ns = this.getStackInSlot( slot );
-				this.inv[slot] = null;
+				this.inv[slot] = AEApi.instance().storage().createItemStack( reachedLimit ? ItemHandlerHelper.copyStackWithSize( stack, limit ) : stack );
 			}
 			else
 			{
-				ns = split.splitStack( qty );
+				existing.grow( reachedLimit ? limit : stack.getCount() );
 			}
-
 			if( this.te != null && Platform.isServer() )
 			{
-				this.te.onChangeInventory( this, slot, InvOperation.decreaseStackSize, ns, ItemStack.EMPTY );
+				this.te.onChangeInventory( this, slot, InvOperation.insert, ItemStack.EMPTY,
+						reachedLimit ? ItemHandlerHelper.copyStackWithSize( stack, limit ) : stack );
 			}
-
-			return ns;
 		}
-
-		return ItemStack.EMPTY;
+		return reachedLimit ? ItemHandlerHelper.copyStackWithSize( stack, stack.getCount() - limit ) : ItemStack.EMPTY;
 	}
 
 	@Override
-	public ItemStack removeStackFromSlot( final int var1 )
+	public ItemStack extractItem( int slot, int amount, boolean simulate )
 	{
+		if( this.inv[slot] != null )
+		{
+			final ItemStack split = this.getStackInSlot( slot );
+
+			if( amount >= split.getCount() )
+			{
+				if( !simulate )
+				{
+					this.inv[slot] = null;
+					if( this.te != null && Platform.isServer() )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)